File: problems-extend-adapt/0201-0300/0220_near-duplicate-in-range-iii/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {

    public boolean hasNearDuplicateInRange(long[] nums, long indexDiff, long valueDiff) {
        // Value buckets of width valueDiff + 1, keyed by floor division: two
        // values in one bucket are within valueDiff by construction, so each
        // bucket holds at most one live value and a same-bucket hit is a "yes".
        Map<Long, Long> buckets = new HashMap<>();
        long width = valueDiff + 1;
        for (int index = 0; index < nums.length; ++index) {
            if (index > indexDiff) {
                // The window spans only the previous indexDiff positions;
                // retire the bucket of the value that just fell out of it.
                buckets.remove(Math.floorDiv(nums[(int) (index - indexDiff - 1)], width));
            }
            long value = nums[index];
            long bucket = Math.floorDiv(value, width);
            if (buckets.containsKey(bucket)) return true;
            // Neighbor buckets can hold values up to 2*valueDiff away, so
            // their occupants need a real distance comparison.
            Long below = buckets.get(bucket - 1);
            if (below != null && value - below <= valueDiff) return true;
            Long above = buckets.get(bucket + 1);
            if (above != null && above - value <= valueDiff) return true;
            buckets.put(bucket, value);
        }
        return false;
    }
}

File: problems-extend-adapt/0201-0300/0220_near-duplicate-in-range-iii/solution.go
// floorDiv divides the way buckets need: Go's / truncates toward zero, and
// negative values must land in the bucket they belong to.
func floorDiv(x, w int64) int64 {
	q := x / w
	if x%w != 0 && x < 0 {
		q--
	}
	return q
}

// Value buckets of width valueDiff + 1, keyed by floor division: two values
// in one bucket are within valueDiff by construction, so each bucket holds at
// most one live value and a same-bucket hit is a "yes".
func hasNearDuplicateInRange(nums []int64, indexDiff int64, valueDiff int64) bool {
	width := valueDiff + 1
	buckets := make(map[int64]int64)
	for index, value := range nums {
		if int64(index) > indexDiff {
			// The window spans only the previous indexDiff positions;
			// retire the bucket of the value that just fell out of it.
			delete(buckets, floorDiv(nums[int64(index)-indexDiff-1], width))
		}
		bucket := floorDiv(value, width)
		if _, ok := buckets[bucket]; ok {
			return true
		}
		// Neighbor buckets can hold values up to 2*valueDiff away, so their
		// occupants need a real distance comparison.
		if below, ok := buckets[bucket-1]; ok && value-below <= valueDiff {
			return true
		}
		if above, ok := buckets[bucket+1]; ok && above-value <= valueDiff {
			return true
		}
		buckets[bucket] = value
	}
	return false
}
